examples/flow.cpp

```diff
@@ -203,45 +273,87 @@ try
-            for (int c = 0; c < nc; ++c) {
-                state.gasoilratio()[c] = state.surfacevol()[c*np + pu.phase_pos[Gas]]
-                    / state.surfacevol()[c*np + pu.phase_pos[Oil]];
+            const int numPhases = props->numPhases();
+            const int numCells  = Opm::UgGridHelpers::numCells(grid);
+            for (int c = 0; c < numCells; ++c) {
+                state.gasoilratio()[c] = state.surfacevol()[c*numPhases + pu.phase_pos[Gas]]
+                    / state.surfacevol()[c*numPhases + pu.phase_pos[Oil]];
             }
         }
     } else if (deck->hasKeyword("EQUIL") && props->numPhases() == 3) {
-        state.init(*grid->c_grid(), props->numPhases());
+        state.init(Opm::UgGridHelpers::numCells(grid),
+                   Opm::UgGridHelpers::numFaces(grid),
+                   props->numPhases());
         const double grav = param.getDefault("gravity", unit::gravity);
-        initStateEquil(*grid->c_grid(), *props, deck, eclipseState, grav, state);
-        state.faceflux().resize(grid->c_grid()->number_of_faces, 0.0);
+        initStateEquil(grid, *props, deck, eclipseState, grav, state);
+        state.faceflux().resize(Opm::UgGridHelpers::numFaces(grid), 0.0);
     } else {
-        initBlackoilStateFromDeck(*grid->c_grid(), *props, deck, gravity[2], state);
+        initBlackoilStateFromDeck(Opm::UgGridHelpers::numCells(grid),
+                                  Opm::UgGridHelpers::globalCell(grid),
+                                  Opm::UgGridHelpers::numFaces(grid),
+                                  Opm::UgGridHelpers::faceCells(grid),
+                                  Opm::UgGridHelpers::beginFaceCentroids(grid),
+                                  Opm::UgGridHelpers::beginCellCentroids(grid),
+                                  Opm::UgGridHelpers::dimensions(grid),
+                                  *props, deck, gravity[2], state);
     }
 
+
     // The capillary pressure is scaled in new_props to match the scaled capillary pressure in props.
     if (deck->hasKeyword("SWATINIT")) {
-        const int nc = grid->c_grid()->number_of_cells;
-        std::vector<int> cells(nc);
-        for (int c = 0; c < nc; ++c) { cells[c] = c; }
+        const int numCells = Opm::UgGridHelpers::numCells(grid);
+        std::vector<int> cells(numCells);
+        for (int c = 0; c < numCells; ++c) { cells[c] = c; }
         std::vector<double> pc = state.saturation();
-        props->capPress(nc, state.saturation().data(), cells.data(), pc.data(),NULL);
+        props->capPress(numCells, state.saturation().data(), cells.data(), pc.data(),NULL);
         new_props->setSwatInitScaling(state.saturation(),pc);
     }
 
     bool use_gravity = (gravity[0] != 0.0 || gravity[1] != 0.0 || gravity[2] != 0.0);
     const double *grav = use_gravity ? &gravity[0] : 0;
 
+    const bool use_local_perm = param.getDefault("use_local_perm", true);
+
+    std::shared_ptr<DerivedGeology> geoprops;
```

Comment:
> I prefer stack objects. 

is there any real-world advantages to this? in this case, it can't be performance...